### PR TITLE
Automatically update map when hosting a game

### LINKED ISSUE #1986: Automatically update map when hosting a game
I've a few maps which got a few plays, one actually got over 3k plays now and I updated the map and added textures and stuff but people still keep playing the old version...

It would be a nice feature to have to either

+force hosts to download the latest version of the map (?)
+or at least display a notice that there is a newer version and that the host can download the map before he creates the lobby

### PR DESCRIPTION
**Pull request by Marc-Spector**
Closes #1986

### CODE REVIEW

**Comment by Marc-Spector** [comment]
Do I need to add unit tests? What do you think?

**Comment by Sheikah45** [comment]
Tests are always good. Helps to make sure that things don't get broken by future changes

File: src/main/java/com/faforever/client/map/MapBean.java [hunk 1]
```diff
@@ -243,6 +243,20 @@ public int compareTo(@NotNull MapBean o) {
...
+
+    ComparableVersion v1 = this.getVersion();
+    ComparableVersion v2 = o.getVersion();
+    if (v1 == null || v2 == null) {
+      throw new CompareMapVersionException(String.format("cannot compare map versions, map '%s' has null version",
+          v1 == null ? getFolderName() : o.getFolderName()));
+    }
+    return v1.compareTo(v2);
+  }
+
```
Comment (Sheikah45) [inline src/main/java/com/faforever/client/map/MapBean.java]: I would move this operation to the MapService as the MapBean is really just a data storage object

Comment (Sheikah45) [reply src/main/java/com/faforever/client/map/MapBean.java]: Also the latestMapVersion should always be the most recent so you don't have to check the version specifically just check if the latest is the same as the one you are comparing to.

Comment (Marc-Spector) [reply src/main/java/com/faforever/client/map/MapBean.java]: Good! I did not think about it.

File: src/main/java/com/faforever/client/map/MapService.java [hunk 2]
```diff
@@ -475,6 +482,60 @@ public void evictCache() {
...
+
+  public CompletableFuture<Optional<MapBean>> getUpdatedMapIfExist(MapBean map) {
+    return getLatestVersionMap(map).thenApply(optional -> {
+      if (optional.isPresent()) {
+        MapBean latestMap = optional.get();
+        try {
+          if (isNewVersionMap(latestMap, map)) {
+            return Optional.of(latestMap);
+          }
+        } catch (CompareMapVersionException ignored) {  }
```
Comment (Sheikah45) [inline src/main/java/com/faforever/client/map/MapService.java]: should at least add a log statement

**Review by Sheikah45 — CHANGES_REQUESTED** [review]
Very thorough tests!
Looks good just a log statement in the ignored exception would be nice

**Review by Sheikah45 — APPROVED** [review]

**Pull request merged**